src/org/traccar/api/AsyncSocketServlet.java

```diff
@@ -22,25 +22,113 @@
 import org.eclipse.jetty.websocket.servlet.WebSocketServletFactory;
 import org.traccar.Context;
 import org.traccar.api.resource.SessionResource;
+import org.traccar.helper.Log;
+import org.traccar.model.User;
+
+import javax.annotation.security.PermitAll;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.container.ResourceInfo;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.SecurityContext;
+import java.lang.reflect.Method;
+import java.sql.SQLException;
+
 
 public class AsyncSocketServlet extends WebSocketServlet {
 
+    public static final String AUTHORIZATION_HEADER = "Authorization";
+    public static final String WWW_AUTHENTICATE = "WWW-Authenticate";
+    public static final String BASIC_REALM = "Basic realm=\"api\"";
+    public static final String X_REQUESTED_WITH = "X-Requested-With";
+    public static final String XML_HTTP_REQUEST = "XMLHttpRequest";
     private static final long ASYNC_TIMEOUT = 10 * 60 * 1000;
 
+    @javax.ws.rs.core.Context
+    private HttpServletRequest request;
+
+    @javax.ws.rs.core.Context
+    private ResourceInfo resourceInfo;
+
+    @javax.ws.rs.core.Context
+    private SecurityContext securityContext;
+
+    protected long getUserId() {
+        UserPrincipal principal = (UserPrincipal) securityContext.getUserPrincipal();
+        if (principal != null) {
+            return principal.getUserId();
+        }
+        return 0;
+    }
+
     @Override
     public void configure(WebSocketServletFactory factory) {
         factory.getPolicy().setIdleTimeout(Context.getConfig().getLong("web.timeout", ASYNC_TIMEOUT));
         factory.setCreator(new WebSocketCreator() {
             @Override
             public Object createWebSocket(ServletUpgradeRequest req, ServletUpgradeResponse resp) {
-                if (req.getSession() != null) {
-                    long userId = (Long) req.getSession().getAttribute(SessionResource.USER_ID_KEY);
-                    return new AsyncSocket(userId);
-                } else {
-                    return null;
-                }
+
+                auth(req);
+                return new AsyncSocket(getUserId());
+
             }
         });
     }
 
+    private void auth(ServletUpgradeRequest req) {
```

Comment:
It looks very similar to the code that we have in `SecurityRequestFilter`. Is there a way to avoid duplication?

Reply:
I do not have much experience with Servlet Filters, so I could not add SecurityRequestFilter to this servlet.